### PR TITLE
CP-921: Support all regions; Remove non-user facing parameters

### PR DESCRIPTION
**Pull request by Tankanow**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -34,3 +34,4 @@ coverage-report
 packaged.template.yaml
 .cz_py_dependencies_installed
 .aws-sam
+app.zip
```
Comment (Tankanow) [inline .gitignore]: You'll see below ... b/c we now need to deploy the Lambda zip files to _every_ region; we don't use `sam package/deploy` directly.

File: Makefile [hunk 2]
```diff
@@ -93,14 +93,14 @@ clean: clean-sam-apps
 #
 ###################
 
-cfn-deploy: guard-stack_name guard-template_file
+cfn-deploy: guard-stack_name guard-template_file guard-bucket
```
Comment (Tankanow) [inline Makefile]: we pass this in now

File: Makefile [hunk 3]
```diff
@@ -149,20 +149,39 @@ deploy-dry-run: $(VIRTUAL_ENV) $(PACKAGED_TEMPLATE_FILE)
 	@$(MAKE) cfn-dryrun stack_name=cz-$(FEATURE_NAME) template_file=$(PACKAGED_TEMPLATE_FILE)
 
 
+.PHONY: deploy-once
```
Comment (Tankanow) [inline Makefile]: `deploy-once` sets up a `cz-provision-account-${region}` bucket in every region and a special `cz-provision-account` bucket in `us-east-1` ... I took this out of `deploy` b/c this should rarely change

File: Makefile [hunk 4]
```diff
@@ -149,20 +149,39 @@ deploy-dry-run: $(VIRTUAL_ENV) $(PACKAGED_TEMPLATE_FILE)
...
+		export AWS_DEFAULT_REGION="$${r}" && \
+		$(MAKE) cfn-deploy stack_name="cz-$(FEATURE_NAME)-$${r}" template_file=$(PACKAGED_TEMPLATE_FILE) bucket="$(BUCKET)-$${r}" ; \
+	done
+
+
 .PHONY: deploy                                                                          ## Deploys Artifacts to S3 Bucket
 deploy:
 	@. ./project.sh && cz_assert_profile && \
-	$(MAKE) $(PACKAGED_TEMPLATE_FILE) && \
+	regions=`aws ec2 describe-regions | jq -r -e '.Regions[].RegionName'` && \
```
Comment (Tankanow) [inline Makefile]: And now we need to manually `aws s3 cp` the zipped applications to each region

File: services/account_type/cloudtrail_owner.yaml [hunk 5]
```diff
@@ -36,39 +36,15 @@ Resources:
         - Key: cloudzero-reactor-account-id
           Value: !Ref ReactorAccountId
 
-  SnsTopicPolicy:
```
Comment (Tankanow) [inline services/account_type/cloudtrail_owner.yaml]: We don't need this Policy anymore, our future Reactor Lambdas can subscribe directly to SQS

File: services/account_type/cloudtrail_owner.yaml [hunk 6]
```diff
@@ -36,39 +36,15 @@ Resources:
...
-            Resource: !Ref CloudTrailSNSTopicArn
-
   SnsSubscription:
     Type: AWS::SNS::Subscription
     Condition: CreateResources
     Properties:
       Protocol: sqs
       Endpoint: !GetAtt SqsQueue.Arn
-      Region: !Ref "AWS::Region"
+      Region: !Select
```
Comment (Tankanow) [inline services/account_type/cloudtrail_owner.yaml]: b/c this SNS Topic could be in another region, read the region from the Topic Arn

File: services/account_type/legacy.yaml [hunk 7]
```diff
@@ -79,12 +79,7 @@ Resources:
     Type: AWS::IAM::Role
     Condition: CreateResources
     Properties:
-      Tags:
```
Comment (Tankanow) [inline services/account_type/legacy.yaml]: Apparently `Role`s are not Taggable in all regions

File: services/connected_account.yaml [hunk 8]
```diff
@@ -9,35 +9,20 @@ Metadata:
   AWS::CloudFormation::Interface:
     ParameterGroups:
       - Label:
-          default: Account Name Metadata for your Account
+          default: Required
```
Comment (Tankanow) [inline services/connected_account.yaml]: Removes most of the parameters, put them in a `Mappings` instead

File: services/discovery/Makefile [hunk 9]
```diff
@@ -104,12 +104,9 @@ $(BUILD_DIR): $(SRC_FILES) $(TEMPLATE_FILE)
 	sam build
 
 .PHONY: package
-package: $(PACKAGED_TEMPLATE_FILE)
-$(PACKAGED_TEMPLATE_FILE): $(BUILD_DIR)
-	@. ../../project.sh && cz_assert_profile && \
-	sam package \
-		--s3-bucket $(bucket) \
-		--output-template-file $@
+package: $(APP_ZIP)
```
Comment (Tankanow) [inline services/discovery/Makefile]: Here we manually zip up the lambda. We can still use `sam build` instead of our vendoring

File: services/discovery/template.yaml [hunk 10]
```diff
@@ -16,8 +16,10 @@ Resources:
   DiscoveryFunction:
     Type: AWS::Serverless::Function 
     Properties:
-      CodeUri: src/
-      Handler: app.handler
+      CodeUri:
```
Comment (Tankanow) [inline services/discovery/template.yaml]: And here we reference the lambda code dynamically in whichever region the stack is being deployed

File: services/notification/Makefile [hunk 11]
```diff
@@ -17,6 +17,17 @@ COVERAGE_REPORT = coverage-report
 FLAKE8_OUT = flake8.out
 BUILD_DIR = .aws-sam
 DEPLOYMENT_FILE = .deploy
+ACCOUNT_ID = $(shell source ../../project.sh && cz_assert_profile &> /dev/null && \
```
Comment (Tankanow) [inline services/notification/Makefile]: copied some missing stuff from another Makefile

File: services/notification/event.json [hunk 12]
```diff
@@ -4,9 +4,9 @@
   "RequestId": "Some Request Id",
   "RequestType": "Create",
   "ResourceProperties": {
-    "ReactorSQSQueueUrl": "str",
+    "ReactorCallbackUrl": "str",
```
Comment (Tankanow) [inline services/notification/event.json]: Using URL not SQS directly

File: services/account_type/legacy.yaml [hunk 13]
```diff
@@ -79,12 +79,7 @@ Resources:
...
-        - Key: cloudzero-stack
-          Value: !Ref AWS::StackName
-        - Key: cloudzero-reactor-account-id
-          Value: !Ref ReactorAccountId
```
Comment (killthrush) [inline services/account_type/legacy.yaml]: Are tags assigned by the stack itself?

File: services/connected_account.yaml (hunk 8 above)
Comment (killthrush) [inline services/connected_account.yaml]: Was this change intentional?  I'm kinda confused by it

File: services/connected_account.yaml [hunk 14]
```diff
@@ -9,35 +9,20 @@ Metadata:
...
-      - Label:
-          default: CloudZero Callback Information
-        Parameters:
-          - ReactorId
-          - ReactorSQSQueueUrl
     ParameterLabels:
       AccountName:
         default: |
-          Optional Name you can set for your AWS Account; you can use this as a descriptive alias for AccountId
+          Descriptive Alias for your AWS Account; Generated by You; Used to ID your accounts in CloudZero platform
```
Comment (killthrush) [inline services/connected_account.yaml]: lowercase "you"?

File: services/connected_account.yaml [hunk 15]
```diff
@@ -52,25 +37,22 @@ Parameters:
...
+Mappings:
+  CallbackConfiguration:
+    dev:
+      ReactorAccountId: '998146006915'
+      ReactorId: '67ad3e83-1ec5-4f4f-b505-c27e2758b990'
+      ReactorCallbackUrl: 'null'
+    prod:
+      ReactorAccountId: '061190967865'
+      ReactorId: 'f7a07d02-d509-4e8d-9fb9-69c7985a6bd8'
+      ReactorCallbackUrl: 'https://api.cloudzero.com/accounts/v1/link'
```
Comment (killthrush) [inline services/connected_account.yaml]: This one concerns me, as we can't cover all of our namespaces with this approach.  Can we convert this to a custom resource at some point?  Then we could resolve the reactor ID via runtime config instead.

File: services/connected_account.yaml [hunk 16]
```diff
@@ -91,14 +73,14 @@ Resources:
...
-          Value: !Ref ReactorAccountId
+          Value: !FindInMap [ CallbackConfiguration, prod, ReactorAccountId ]
       Parameters:
         IsResourceOwnerAccount: !GetAtt Discovery.Outputs.IsResourceOwnerAccount
         IsAuditAccount: !GetAtt Discovery.Outputs.IsAuditAccount
         IsMasterPayerAccount: !GetAtt Discovery.Outputs.IsMasterPayerAccount
         IsCloudTrailOwnerAccount: !GetAtt Discovery.Outputs.IsCloudTrailOwnerAccount
         ExternalId: !Ref ExternalId
-        ReactorAccountId: !Ref ReactorAccountId
+        ReactorAccountId: !FindInMap [ CallbackConfiguration, prod, ReactorAccountId ]
```
Comment (killthrush) [inline services/connected_account.yaml]: fine for initial testing, not worth blocking the PR but we should discuss how to improve this for later

File: services/discovery/Makefile (hunk 9 above)
Comment (killthrush) [reply services/discovery/Makefile]: Is the packaged template file handled by `sam build`?

File: services/discovery/template.yaml [hunk 17]
```diff
@@ -16,8 +16,10 @@ Resources:
...
+        Bucket: !Sub 'cz-provision-account-${AWS::Region}'
```
Comment (killthrush) [inline services/discovery/template.yaml]: Is this our bucket?  Which account is it in?  We can only have one of these in each region without having name collisions.

File: services/account_type/legacy.yaml (hunk 13 above)
Comment (Tankanow) [reply services/account_type/legacy.yaml]: The stacks _do_ have tags. It does not look like the `Role`s have tags =(. They are under the `/cloudzero/` `Path` at least.

Comment (Tankanow) [reply services/account_type/legacy.yaml]: It's very weird, b/c this `Tags:` CFN property _works_ in `us-east-1` but not in _all_ other regions I've tried.

File: services/connected_account.yaml (hunk 8 above)
Comment (Tankanow) [reply services/connected_account.yaml]: Yes. We wanted to make it clearer to the user which CFN parameters are required. This section is for the Labels that appear in the CFN "Create Stack" Console

File: services/connected_account.yaml (hunk 14 above)
Comment (Tankanow) [reply services/connected_account.yaml]: fixed

File: services/discovery/Makefile (hunk 9 above)
Comment (Tankanow) [reply services/discovery/Makefile]: We actually _don't_ want the packaged template b/c `sam package` replaces the `CodeUri` with its own.

File: services/discovery/template.yaml (hunk 17 above)
Comment (Tankanow) [reply services/discovery/template.yaml]: These are our buckets. They are in `cz-prime`. It sucks to have to create ~ 20 buckets; but it's the only way this will work (without inline CFN Lambdas ... [shudder]).

File: services/connected_account.yaml (hunk 15 above)
Comment (Tankanow) [reply services/connected_account.yaml]: I'll add a TODO

Comment (Tankanow) [reply services/connected_account.yaml]: Remember, these resources get deployed in our customers' accounts, where namespaces don't really apply.

File: services/connected_account.yaml (hunk 16 above)
Comment (Tankanow) [reply services/connected_account.yaml]: agreed

File: services/connected_account.yaml (hunk 15 above)
Comment (killthrush) [reply services/connected_account.yaml]: right, what I means was that each of those reactor IDs maps to one of our namespaces.

Comment (Tankanow) [reply services/connected_account.yaml]: Right right ... that makes sense.

File: services/discovery/template.yaml (hunk 17 above)
Comment (killthrush) [reply services/discovery/template.yaml]: OK, as long as we don't get collisions that will be fine 👍
